### PR TITLE
Audit: Fix issue #38

### PR DESCRIPTION
**Pull request by sbuljac**
/sherlock-audit/2023-12-notional-update-5-judging/issues/38

### CODE REVIEW

File: contracts/internal/balances/ExternalLending.sol [hunk 1]
```diff
@@ -78,10 +78,16 @@ library ExternalLending {
...
+            // L as the currentExternalUnderlyingLend
+            // W as the externalUnderlyingAvailableForWithdraw
+            // D as the Deposit (the variable we want to solve for)
+            // T / 100 = (W + D) / (L + D)
+            // D = (100 * W - L * T) / (T - 100)
+            maxExternalUnderlyingLend = oracleData.externalUnderlyingAvailableForWithdraw
                 .mul(uint256(Constants.PERCENTAGE_DECIMALS))
-                .div(rebalancingTargetData.externalWithdrawThreshold);
+                .sub(oracleData.currentExternalUnderlyingLend.mul(rebalancingTargetData.externalWithdrawThreshold))
+                .div(uint256(rebalancingTargetData.externalWithdrawThreshold - Constants.PERCENTAGE_DECIMALS));
```
Comment (jeffywu) [inline contracts/internal/balances/ExternalLending.sol:90]: Add a comment that underflow is checked when value is set

File: contracts/external/actions/TreasuryAction.sol [hunk 2]
```diff
@@ -124,7 +124,7 @@ contract TreasuryAction is StorageLayoutV2, ActionGuards, NotionalTreasury {
 
         require(config.holding == holding);
         require(config.targetUtilization < 100);
-        require(100 <= config.externalWithdrawThreshold);
+        require(100 < config.externalWithdrawThreshold);
```
Comment (jeffywu) [inline contracts/external/actions/TreasuryAction.sol:127]: @T-Woodward should this always be greater than 100?

Comment (T-Woodward) [reply contracts/external/actions/TreasuryAction.sol:127]: Not necessarily. Again, could be something like sDAI where there is no redemption/illiquidity risk. In that case you wouldn't need a withdrawal buffer theoretically

File: contracts/internal/balances/ExternalLending.sol (hunk 1 above)
Comment (jeffywu) [reply contracts/internal/balances/ExternalLending.sol:90]: @T-Woodward this math prevents setting the threshold at exactly 100, but in theory it should work for greater than or less than zero values. I think the math would make more sense to me if we reversed the signs in the numerator and denominator:

`// D = (L * T - W * 100) / (100 - T)`

Although this is equivalent.

Comment (T-Woodward) [reply contracts/internal/balances/ExternalLending.sol:90]: Ok. In practice, it would never be necessary to set T to exactly 100 so we can require that not ever happen and it will be fine. As for the math, I find both expressions equally unintuitive though I do think they're correct

Comment (jeffywu) [reply contracts/internal/balances/ExternalLending.sol:90]: Ok, so looking through the code this is what I think:

In this if block we basically do not allow the withdrawThreshold to go below zero since we would get a sub underflow and other weird stuff would happen:
https://github.com/notional-finance/contracts-v2/blob/68fb703feed15dcc27abad91d97780203ba70694/contracts/internal/balances/ExternalLending.sol#L80

@sbuljac is right, if the value is below zero then we have a problem because we may try to deposit while there is some sort of bank run going on in Aave.

I think for something like sDAI we should allow the withdrawThreshold = 100. That basically means that we expect to always be able to withdraw our funds at a 1-1 basis and we do not require overhead to ensure that we can withdraw before someone else does.

When it comes to this equation, I think we can have the special case where withdrawThreshold = 100 then the maxExternalUnderlyingLend = type(uint256).max.

Then in this line:
https://github.com/notional-finance/contracts-v2/blob/68fb703feed15dcc27abad91d97780203ba70694/contracts/internal/balances/ExternalLending.sol#L89

We will always go down to the min value which will be:
`uint256(underlyingToken.convertToExternal(targetExternalUnderlyingLend))`

Basically we are not going to cap our deposits based on what the external market has, we assume that we can always deposit up to our target utilization (this is all in the case where withdrawThreshold = 100).

In practice, I'm not sure if we would rather just set the threshold to 101 and ensure that there is some bit of buffer, but given the nature of sDAI this would only ever be a problem if we were the last ones out of the contract so basically irrelevant.

Comment (T-Woodward) [reply contracts/internal/balances/ExternalLending.sol:91]: Ok either way works for me
